### PR TITLE
[BUG] Applying patch cupy 3218 to build to get full CUDA 10.2 support.

### PR DESCRIPTION
**Pull request by rlratzel**
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Fixes error exclusive to CUDA 10.2 environments where cupy raises an error:
```
catastrophic error: cannot open source file "cuda_fp16.h"
```
This PR cherry-picks commits to add full CUDA 10.2 support, which was added in cupy commit https://github.com/cupy/cupy/commit/d654524fa1c5aae4df39090d148f5eadce4ebb2c

The cupy package was built locally and installed in a CUDA 10.2 environment, where the previously-failing test then passed.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -28,6 +27,9 @@ build:
     - export CXXFLAGS="${CXXFLAGS} -I/usr/include"
     #- export CUB_PATH=${PWD}/cub_v1.8.0  # TODO(leofang): remove this
     - "{{ PYTHON }} -m pip install . -vv"
+    # full CUDA 10.2 support was added in a commit post-7.2.0 release
+    # (d654524fa), cherry pick it and a prior commit (34d4dbd73) to support it
+    - git cherry-pick --strategy=recursive -Xtheirs -n 34d4dbd73eb534217095273499bfde03dc508d47 d654524fa1c5aae4df39090d148f5eadce4ebb2c
```
Comment (jakirkham) [inline recipe/meta.yaml]: To get patches from these, can just pull them from GitHub like so. Would add those here and then we can still use the tarball.

https://github.com/cupy/cupy/commit/34d4dbd73eb534217095273499bfde03dc508d47.patch
https://github.com/cupy/cupy/commit/d654524fa1c5aae4df39090d148f5eadce4ebb2c.patch

**Comment by leofang** [comment]
@rlratzel Could you please wait for a little while and target v7.3.0, which just gets released a few hours ago? The bot will generate a PR, and you can rebase on top of that.

Also, please follow John's comment on the patching. I'd suggest you to apply the patch from cupy/cupy#2910 directly, instead of doing a git cherry-pick. It's nerve wracking to me as it'd be hard to maintain and keep track...

@kmaehashi Any chance you could do the backport of cupy/cupy#2910 asap?

**Comment by leofang** [comment]
@jakirkham Another thing that has been haunting us for quite a while: Do we want to do our own patches in the absence of CUDA-enabled CIs? This always makes me nervous...I prefer (as always) the burden to be on the PFN-CuPy team, who has the resource and bandwidth to do fixes properly and permanently, not on conda-forge maintainers...

This is not my hard "NO" though, just wondering what you think and what other popular recipes do.

**Comment by leofang** [comment]
> Also, please follow John's comment on the patching. I'd suggest you to apply the patch from [cupy/cupy#2910](https://github.com/cupy/cupy/pull/2910) directly, instead of doing a git cherry-pick. It's nerve wracking to me as it'd be hard to maintain and keep track...

See how we did a patch for CUB earlier: #32.

**Comment by jakirkham** [comment]
The commits actually don't appear to apply cleanly unfortunately. Since it sounds like there is a new CuPy release that is imminent, maybe we should just to try get a patch in there. I've taken a pass at doing this in PR ( https://github.com/cupy/cupy/pull/3218 ). @rlratzel @leofang, could you please check that my patch looks ok? 🙂

**Comment by jakirkham** [comment]
> Could you please wait for a little while and target v7.3.0, which just gets released a few hours ago? The bot will generate a PR, and you can rebase on top of that.

Looks like v7.3.0 was already cut ( https://github.com/cupy/cupy/releases/tag/v7.3.0 ).

**Comment by jakirkham** [comment]
Leo, for context @rlratzel is a colleague of mine. We ran into a CUDA 10.2 issue in CuPy that was impacting cuML. So he was trying to help out here in applying that fix.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -9,16 +9,19 @@ package:
...
+  # CUB support has too many bugs, let us wait for a bit longer
   ## 1. The CUB support has been extensively tested with v1.8.0, so we should pin it here
   ## 2. This dependency should be removed once CuPy bundles CUB (cupy/cupy#2584)
   #- url: https://github.com/NVlabs/cub/archive/v1.8.0.tar.gz
   #  sha256: 025658f4c933cd2aa8cc88a559d013338d68de3fa639cc1f2b12cf61dc759667
   #  folder: cub_v1.8.0  # so the source code is in cupy/cub_v1.8.0
 
 build:
-  number: 1
+  number: 2
```
Comment (leofang) [inline recipe/meta.yaml]: I am going to modify this single line shortly, as we just bumped to 2 in #42 and I am not sure why this doesn't cause conflict. 
```suggestion
  number: 3
```

**Comment by leofang** [comment]
@conda-forge-admin, please rerender

**Comment by leofang** [comment]
> We ran into a CUDA 10.2 issue in CuPy that was impacting cuML. So he was trying to help out here in applying that fix.

Thanks for filling me in, John! Sorry to both of you if my reaction was perceived unfriendly to you. I didn't mean that...

**Comment by leofang** [comment]
Once the CI passes, I'll merge and then update the recipe.

**Comment by rlratzel** [comment]
> Sorry to both of you if my reaction was perceived unfriendly to you. I didn't mean that...

No worries at all, thanks for your guidance.

File: recipe/meta.yaml (hunk 2 above)
Comment (leofang) [reply recipe/meta.yaml]: OK after committing this the conflict appeared...Resolved in 87a95d5. (Sorry @rlratzel for doing the rebase on your behalf 😅)

**Comment by leofang** [comment]
All CI passes. Thanks again for help, Rick! Merging...

**Comment by jakirkham** [comment]
Thanks Leo! 😀

**Comment by jakirkham** [comment]
> Sorry to both of you if my reaction was perceived unfriendly to you. I didn't mean that...

Not at all. The points you raised seemed valid.

Still mulling over how we can improve testing around CuPy changes.

**Pull request merged**